File: programs/gfx-ssl/src/states/ssl.rs
use std::io::Write;
use crate::utils::PDAIdentifier;
use anchor_lang::prelude::*;
use std::mem;
use anchor_lang::Discriminator;

impl PDAIdentifier for SSL {
    const IDENT: &'static [u8] = b"GFX-SSL";

    #[inline(always)]
    fn program_id() -> &'static Pubkey {
        &crate::ID
    }
}

#[account(zero_copy)]
#[cfg_attr(feature = "no-entrypoint", derive(Debug))]
pub struct SSL {
    pub controller: Pubkey,
    pub mint: Pubkey,
    pub decimals: u8, // a copy from the mint
    pub bump: u8,
    pub pt_bump: u8,

    // configurable settings
    pub suspended: u8,
    pub cranker: Pubkey, // the pubkey for the cranker
    pub _pad0: [u8; 4],

    // pool states
    pub weight: u64, // the connector token amount. This does not need to be an SPL token because this is internal
    pub liability: u64, // liability to the user
    pub swapped_liability: u64, // liability that in the form of other tokens due to swapped away, updated by the cranker
    pub total_share: u64,

    pub _pad1: [u64; 32],
}

const _: [u8; 392] = [0; std::mem::size_of::<SSL>()];

impl Default for SSL {
    fn default() -> Self {
        unsafe { mem::zeroed() }
    }
}

impl AccountSerialize for SSL {
    fn try_serialize<W: Write>(&self, writer: &mut W) -> Result<()> {
        let mut disc = Self::discriminator().to_vec();
        disc.append(&mut bytemuck::bytes_of(self).to_vec());
        writer.write_all(&disc)?;
        Ok(())
    }
}

File: ts/src/states/ssl.ts
import * as lo from '@solana/buffer-layout';
import { PublicKey } from "@solana/web3.js";
import { publicKey, u64, bool } from "@solana/buffer-layout-utils";

export interface SSL {
    sighash: Uint8Array,
    controller: PublicKey,
    mint: PublicKey,
    decimals: number,
    bump: number,
    ptBump: number,
    suspended: boolean,
    cranker: PublicKey,
    weight: BigInt,
    liablity: BigInt,
    swappedLiability: BigInt,
    totalShare: BigInt,
}

export const SSL_LAYOUT = lo.struct<SSL>([
    lo.blob(8, "sighash"),
    publicKey("controller"),
    publicKey("mint"),
    lo.u8("decimals"),
    lo.u8("bump"),
    lo.u8("ptBump"),
    bool("suspended"),
    publicKey("cranker"),
    lo.blob(4), // padding
    u64("weight"),
    u64("liability"),
    u64("swappedLiability"),
    u64("totalShare"),

    lo.blob(256, "padding")
]);

if (SSL_LAYOUT.span != 392 + 8) {
    throw new Error("SSL LAYOUT SIZE MISMATCH");
}